cmd/srcd/cmd/components_test.go

```diff
@@ -0,0 +1,161 @@
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func (s *ComponentsTestSuite) TearDownTest() {
+	s.RunStop(context.Background())
+	os.RemoveAll(s.testDir)
+}
+
+func (s *ComponentsTestSuite) TestListStopped() {
+	require := s.Require()
+
+	out, err := s.RunCommand(context.TODO(), "components", "list")
+	require.NoError(err, out.String())
+
+	expected := regexp.MustCompile(
+		`^IMAGE +INSTALLED +RUNNING +CONTAINER NAME
+bblfsh/bblfshd:\S+ +(yes|no) +no +srcd-cli-bblfshd
+bblfsh/web:\S+ +(yes|no) +no +srcd-cli-bblfsh-web
+srcd/cli-daemon:\S+ +(yes|no) +no +srcd-cli-daemon
+srcd/gitbase-web:\S+ +(yes|no) +no +srcd-cli-gitbase-web
+srcd/gitbase:\S+ +(yes|no) +no +srcd-cli-gitbase
+$`)
+
+	s.Regexp(expected, out.String())
+}
+
+func (s *ComponentsTestSuite) TestListInit() {
+	require := s.Require()
+
+	out, err := s.RunInit(context.TODO(), s.testDir)
+	require.NoError(err, out.String())
+
+	out, err = s.RunCommand(context.TODO(), "components", "list")
+	require.NoError(err, out.String())
+
+	expected := regexp.MustCompile(`srcd/cli-daemon:\S+ +yes +yes +srcd-cli-daemon`)
+	s.Regexp(expected, out.String())
+}
+
+func (s *ComponentsTestSuite) TestInstall() {
+	require := s.Require()
+
+	out, err := s.RunCommand(context.TODO(), "components", "list")
+	require.NoError(err, out.String())
+
+	// Get the exact image:version of gitbase
+	exp := regexp.MustCompile(`(srcd/gitbase:\S+) +(yes|no)`)
+	matches := exp.FindStringSubmatch(out.String())
+
+	require.NotNil(matches)
+	require.Len(matches, 3)
+
+	imgVersion := matches[1]
+	installed := matches[2]
+
+	// If installed, remove it
+	if installed == "yes" {
+		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+		defer cancel()
+		err = docker.RemoveImage(ctx, imgVersion)
+		require.NoError(err)
+	}
+
+	// Check it's not installed
+	out, err = s.RunCommand(context.TODO(), "components", "list")
+	require.NoError(err, out.String())
+
+	expected := regexp.MustCompile(`srcd/gitbase:\S+ +no +no +srcd-cli-gitbase`)
+	require.Regexp(expected, out.String())
+
+	// Install
+	out, err = s.RunCommand(context.TODO(), "components", "install", "srcd/gitbase")
+	require.NoError(err, out.String())
+
+	// Check it's installed
+	out, err = s.RunCommand(context.TODO(), "components", "list")
+	require.NoError(err, out.String())
+
+	expected = regexp.MustCompile(`srcd/gitbase:\S+ +yes +no +srcd-cli-gitbase`)
+	require.Regexp(expected, out.String())
+
+	// Call install again, should be an exit 0
+	out, err = s.RunCommand(context.TODO(), "components", "install", "srcd/gitbase")
+	require.NoError(err, out.String())
+}
+
+func (s *ComponentsTestSuite) TestInstallAlias() {
+	require := s.Require()
+
+	// Install with image name
+	out, err := s.RunCommand(context.TODO(), "components", "install", "srcd/cli-daemon")
+	require.NoError(err, out.String())
+
+	// Install with container name
+	out, err = s.RunCommand(context.TODO(), "components", "install", "srcd-cli-daemon")
```

Comment:
shouldn't we ensure that the required container changed from "not installed" to "installed"?

Reply:
That is already done in `TestInstall()`. I thought in this test it was enough to test that the install command has an exit 0.

We could do the same steps here again: uninstall the image, call `components install`, and check the image was installed. But then it would make the tests too slow, we would be downloading the image 3 times instead of 1. I'd prefer to skip actual image downloads to a minimum.